db/macc_glue.c

```diff
@@ -460,90 +455,156 @@ static int create_key_schema(dbtable *db, struct schema *schema, int alt,
+        return -1;
+    }
+
+    for (ix = 0; ix < sch->nix; ix++) {
+        sch->ix[ix] = s = _create_index_schema(sch->tag, ix, err);
+        if (!s)
+            return -1;
+
+        if (dyns_is_idx_partial_datacopy(ix)) {
+            s->flags |= SCHEMA_PARTIALDATACOPY;
+            s->partial_datacopy = _create_index_datacopy_schema(sch, ix, err);
+            if (!s->partial_datacopy)
+                return -1;
+            tbl->ix_datacopylen[ix] =
+                s->partial_datacopy->member[s->partial_datacopy->nmembers-1].offset +
+                s->partial_datacopy->member[s->partial_datacopy->nmembers-1].len;
+        } else
+            tbl->ix_datacopylen[ix] = 0;
+
+        if (is_ondisk)
+            tbl->ixschema[ix] = sch->ix[ix];
+
+        offset = 0;
+        for (piece = 0; piece < s->nmembers; piece++) {
+            if (_set_schema_index_column(sch, &s->member[piece], ix, piece, offset, err))
+                return -1;
+            offset += s->member[piece].len;
+
             /* we could check here if there are decimals in the index, and mark
              * it "datacopy" special */
-            if (db->ix_datacopy[ix] == 0 && m->type == SERVER_DECIMAL) {
-                db->ix_collattr[ix]++; /* special tail */
-            } else if (m->type == SERVER_DECIMAL) {
-                db->ix_datacopy[ix]++; /* special tail */
+            if (tbl->ix_datacopy[ix] == 0 && s->member[piece].type == SERVER_DECIMAL) {
+                tbl->ix_collattr[ix]++; /* special tail */
+            } else if (s->member[piece].type == SERVER_DECIMAL) {
+                tbl->ix_datacopy[ix]++; /* special tail */
+            }
+            if (!tbl->ix_expr) {
+                if (gbl_new_indexes &&
+                        strncasecmp(tblname, "sqlite_stat", 11) &&
+                        strcasecmp(tblname, "comdb2_oplog") &&
+                        strcasecmp(tblname, "comdb2_commit_log") &&
+                        (!gbl_replicate_local ||
+                         strcasecmp(s->member[piece].name, "comdb2_seqno"))) {
+                    s->member[piece].isExpr = 1;
+                    tbl->ix_partial = 1;
+                }
+                tbl->ix_expr = s->member[piece].isExpr;
             }
-            m->offset = offset;
-            offset += m->len;
         }
-        /* rest of fields irrelevant for indexes */
-        add_tag_schema(dbname, s);
+        tbl->ix_keylen[ix] = offset;
+
+        lst[(*pnlst)++] = s;
+
         rc = dyns_get_idx_tag(ix, altname, MAXTAGLEN, &where);
-        if (rc == 0 && (strcasecmp(schema->tag, ".ONDISK") == 0 ||
-                        strcasecmp(schema->tag, ".NEW..ONDISK") == 0)) {
+        if (!rc && is_ondisk) {
+            struct schema *s_alias;
             if (alt == 0) {
-                add_tag_alias(dbname, s, altname, schema->nmembers);
+                s_alias = clone_schema_index(s, altname, s->nmembers);
```

Comment:
third argument should be sch->nmembers I think. Assuming sch is the table schema and s is an index schema

Reply:
oh yes, this is the datacopy length for the index, and here I think we create a new tagname for it..

Reply:
renamed and fixed